### ISSUE TITLE
Performance buffers vs malloc_shared

### ISSUE THREAD

**Issue by csccva** [AdaptiveCpp/AdaptiveCpp#826]
Hello,

I wonder if I am doing something  wrong or is it a SYCL code. I have a code doing Jacobi iterations. In one version I am declaring two buffers before starting the iteration loop, while in the other version I allocate the variables using `malloc_shared`. In both cases the initialization is done on the cpu. 
The case with buffers seems to be significantly faster than case with USM. My guess is that the difference comes from the cpu-gpu transfers, but as far as I understand the behaviour of the `malloc_shared` there should be only one copy at the beginning of the main loop. 
I am using hipSYCL from development branch, with llvm14 running on top of cuda 11.5 on A100 gpus, using RH 8.

Bellow I pasted the two codes:
First the versions with buffers:
```
//==============================================================
// Matrix Multiplication: DPC++ Basic Parallel Kernel
//==============================================================
// Copyright © 2021 Intel Corporation
//
// SPDX-License-Identifier: MIT
// =============================================================
// Also based on code from https://github.com/ENCCS/OpenACC-CUDA-beginners/blob/main/content/2.02_cuda-heat-equation.rst

#include <chrono>
#include <cstdio>
#include <iostream>
#include <vector>

#include <chrono>
#include <getopt.h>
#include <sycl/sycl.hpp>
using namespace sycl;

int main(int argc, char *argv[]) {

    size_t N = 1024;
    size_t M = 8;
    int VERIFY = 0;
    int PRINT_OUTPUT_MATRIX = 0;

    int arg;
    while ((arg = getopt (argc, argv, "n:m:vp")) != -1)
        switch (arg){
            case 'n':
                N = std::atoi(optarg);
                break;
            case 'm':
                M = std::atoi(optarg);
                break;
            case 'v':
                VERIFY = 1;
                break;
            case 'p':
                PRINT_OUTPUT_MATRIX = 1;
                break;
            case 'h':
                std::cout << std::endl;
                std::cout << "Usage   : ./a.out -n <MATRIX_SIZE> -m <WORK_GROUP_SIZE> -v -p\n\n";
                std::cout << "          [-n] size for matrix, eg: 1024\n";
                std::cout << "          [-m] size of work_group, eg: 8/16\n";
                std::cout << "          [-v] verify output with linear computation on cpu\n";
                std::cout << "          [-p] print output matrix\n";
                std::cout << "Example : ./a.out -n 1024 -m 16 -v -p\n\n";
                std::exit(0);
        }

    //# Define vectors for matricies
    const int nx=N, ny=N;
    const int mx=M, my=M;
    const int t_x=mx+2, t_y=my+2;
    const int niter=100;
    const double factor =0.25;
    std::vector<float> matrix_u(nx*ny);
    std::vector<float> matrix_unew(nx*ny);
    
    range<1> work_items { N*N};

    // Initialize u
  for (int i = 0; i < nx; i++) {
    for (int j = 0; j < ny; j++) {
      int ind = i * ny + j;
      matrix_u[ind] = ((i - nx / 2) * (i - nx / 2)) / nx +
               ((j - ny / 2) * (j - ny / 2)) / ny;
    }
  }

    auto start = std::chrono::high_resolution_clock::now().time_since_epoch().count();

    //# Define queue with default device for offloading computation
    sycl::property_list q_prof{property::queue::enable_profiling{}, sycl::property::queue::in_order{}};
    //queue q{property::queue::enable_profiling{}};
    queue q{default_selector{},q_prof};
    //queue q{property::queue::enable_profiling{}};
    //queue q();
    //queue q{property::queue::enable_profiling{}, property::queue::in_order()};
    event e;
    std::cout << "Offload Device        : " << q.get_device().get_info<info::device::name>() << "\n";
    std::cout << "max_work_group_size   : " << q.get_device().get_info<info::device::max_work_group_size>() << "\n";
    std::cout << "Configuration         : MATRIX_SIZE= " << nx << "x" << ny << "\n";
    auto kernel_duration=0.0;
    std::cout << " [0][0] = " << matrix_u[0] << "\n";
        //# Create buffers for matrices
    {   
        /*buffer u(matrix_u);
        buffer unew(matrix_unew);*/

        buffer<float> u(matrix_u.data(), work_items);
        buffer<float> unew(matrix_unew.data(), work_items);
        
        for(int iter=0;iter<niter; iter++){ 

            //# Submit command groups to execute on device
            e = q.submit([&](handler &h){
                //# Create accessors to copy buffers to the device
                auto U = u.get_access<access::mode::read_write>(h);
                auto UNEW= unew.get_access<access::mode::read_write>(h);
                
                //# Define size for ND-Range and work-group size
                range<2> global_size(nx,ny);
                range<2> work_group_size(mx,my);
                
                
                //# Create local accessors. They use the memory closer to the chip.
                //# In SYCL called local memory. On nvidia and AMD thw so-called shared memory

                //__shared__ float s_Un[(BLOCK_SIZE_X + 2)*(BLOCK_SIZE_Y + 2)];
                accessor<float, 2, access::mode::read_write, access::target::local> A_tile(range<2>(t_x, t_y), h);

                //# Parallel Compute Matrix Multiplication
                
                h.parallel_for(nd_range<2>{global_size, work_group_size}, [=](nd_item<2> item){
                    const int i = item.get_global_id(0); 
                    const int j = item.get_global_id(1);
                    const int s_i = item.get_local_id(0)+1;
                    const int s_j = item.get_local_id(1)+1;

                    int ind = i * ny + j;
                    int ip = (i + 1) * ny + j;
                    int im = (i - 1) * ny + j;
                    int jp = i * ny + j + 1;
                    int jm = i * ny + j - 1;
                    //Load the data into the local memory (shared memory in CUDA/HIP)
                    //Central square 
                    A_tile[s_i][s_j]=U[ind];
                    //Top border
                    if(s_i==1  && i>0) // (s_i==1 && s_j!=1 && i!=0)?????
                    {
                        A_tile[0][s_j]=U[im]; 
                    }
                    //Bottom border
                    if(s_i==mx  && i<nx-1) // (s_i==mx && s_j1!=my && i<nx-1)
                    {
                        A_tile[mx+1][s_j]=U[ip];
                    } 
                    //Left border
                    if( s_j == 1 && j > 0) // (s_i != 1 && s_j == 1 && j != 0) 
                    {
                        A_tile[s_i][0]=U[jm]; 
                    } 
                    //Right border
                    if( s_j == my && j < ny-1) // (s_i != BLOCK_SIZE_X && s_j == BLOCK_SIZE_Y && j != ny - 1)
                    {
                        A_tile[s_i][my+1]=U[jp]; 
                    } 
                    // missing the corners. They are needed  only if using next nearest neighbors
                    item.barrier(access::fence_space::local_space); // barrier within the group //__syncthreads();
                    if(i>0 && i<nx-1 && j>0 && j< ny-1){
                        UNEW[ind] = factor * (A_tile[s_i-1][s_j]+A_tile[s_i-1][s_j]-2*A_tile[s_i][s_j])
                                  + factor * (A_tile[s_i][s_j+1]+A_tile[s_i][s_j-1]-2*A_tile[s_i][s_j]);
                    }
                });
            });        
            q.wait();
            kernel_duration += (e.get_profiling_info<info::event_profiling::command_end>() - e.get_profiling_info<info::event_profiling::command_start>());
            //std::swap(u, unew);
           }
    }
    q.wait();
    /*for (int i=1; i<nx-1; i++){
            for (int j=1; j<ny-1; j++){
                if(matrix_unew[i*ny+j]>=mx+1 || matrix_unew[i*ny+j]<=0 ){
                    std::cout << matrix_unew[i*ny+j]<< " "<< matrix_u[i*ny+j] << " "  << i << " "<< " " << j << "\n";
                }
                
            }
    }*/
    std::cout << "Kernel Execution Time : " << kernel_duration / 1e+9 << " seconds" << "\n";

    auto duration = std::chrono::high_resolution_clock::now().time_since_epoch().count() - start;
    std::cout << "Compute Duration      : " << duration / 1e+9 << " seconds\n";
}
``` 
The USM based code:

```
#include <chrono>
#include <cstdio>
#include <iostream>
#include <vector>

#include <chrono>
#include <getopt.h>
#include <sycl/sycl.hpp>

using namespace sycl;

int main(int argc, char *argv[]) {

    size_t N = 1024;
    size_t M = 8;
    int VERIFY = 0;
    int PRINT_OUTPUT_MATRIX = 0;

    int arg;
    while ((arg = getopt (argc, argv, "n:m:vp")) != -1)
        switch (arg){
            case 'n':
                N = std::atoi(optarg);
                break;
            case 'm':
                M = std::atoi(optarg);
                break;
            case 'v':
                VERIFY = 1;
                break;
            case 'p':
                PRINT_OUTPUT_MATRIX = 1;
                break;
            case 'h':
                std::cout << std::endl;
                std::cout << "Usage   : ./a.out -n <MATRIX_SIZE> -m <WORK_GROUP_SIZE> -v -p\n\n";
                std::cout << "          [-n] size for matrix, eg: 1024\n";
                std::cout << "          [-m] size of work_group, eg: 8/16\n";
                std::cout << "          [-v] verify output with linear computation on cpu\n";
                std::cout << "          [-p] print output matrix\n";
                std::cout << "Example : ./a.out -n 1024 -m 16 -v -p\n\n";
                std::exit(0);
        }

    //# Define vectors for matrices
    const int nx=N, ny=N;
    const int mx=M, my=M;
    const int t_x=mx+2, t_y=my+2;
    const int niter=500;
    const double factor =0.5;
    auto start = std::chrono::high_resolution_clock::now().time_since_epoch().count();

    //# Define queue with default device for offloading computation
    sycl::property_list q_prof{property::queue::enable_profiling{}, sycl::property::queue::in_order{}};
    //queue q{property::queue::enable_profiling{}};
    queue q{default_selector{},q_prof};
    //queue q{property::queue::enable_profiling{}};
    //queue q();
    //queue q{property::queue::enable_profiling{}, property::queue::in_order()};


    auto matrix_u=malloc_shared<float>(nx*ny, q);
    auto matrix_unew=malloc_shared<float>(nx*ny, q);
    
    range<1> work_items { N*N};
    
    

    // Initialize u
  for (int i = 0; i < nx; i++) {
    for (int j = 0; j < ny; j++) {
      int ind = i * ny + j;
      matrix_u[ind] = ((i - nx / 2) * (i - nx / 2)) / nx +
               ((j - ny / 2) * (j - ny / 2)) / ny;
    }
  }

    event e;
    std::cout << "Offload Device        : " << q.get_device().get_info<info::device::name>() << "\n";
    std::cout << "max_work_group_size   : " << q.get_device().get_info<info::device::max_work_group_size>() << "\n";
    std::cout << "Configuration         : MATRIX_SIZE= " << nx << "x" << ny << "\n";
    auto kernel_duration=0.0;
    std::cout << " [0][0] = " << matrix_u[0] << "\n";
    for(int iter=0;iter<niter; iter++){
        e = q.submit([&](handler &h){
            
            //# Define size for ND-Range and work-group size
            range<2> global_size(nx,ny);
            range<2> work_group_size(mx,my);
            
            //# Create local accessors. They use the memory closer to the chip.
            //# In SYCL called local memory. On nvidia and AMD thw so-called shared memory
            //__shared__ float s_Un[(BLOCK_SIZE_X + 2)*(BLOCK_SIZE_Y + 2)];
            accessor<float, 2, access::mode::read_write, access::target::local> A_tile(range<2>(t_x, t_y), h);

            //# Parallel Compute Matrix Multiplication
            h.parallel_for(nd_range<2>{global_size, work_group_size}, [=](nd_item<2> item){
                const int i = item.get_global_id(0); 
                const int j = item.get_global_id(1);
                const int s_i = item.get_local_id(0)+1;
                const int s_j = item.get_local_id(1)+1;
                
                int ind = i * ny + j;
                int ip = (i + 1) * ny + j;
                int im = (i - 1) * ny + j;
                int jp = i * ny + j + 1;
                int jm = i * ny + j - 1;
                //Load the data into the local memory (shared memory in CUDA/HIP)
                //Central square 
                 A_tile[s_i][s_j]=matrix_u[ind];
                 //Top border
                 if(s_i==1  && i>0) // (s_i==1 && s_j!=1 && i!=0)?????
                 {
                    A_tile[0][s_j]=matrix_u[im]; 
                }
                //Bottom border
                if(s_i==mx  && i<nx-1) // (s_i==mx && s_j1!=my && i<nx-1)
                {
                    A_tile[mx+1][s_j]=matrix_u[ip];
                } 
                //Left border
                if( s_j == 1 && j > 0) // (s_i != 1 && s_j == 1 && j != 0) 
                {
                    A_tile[s_i][0]=matrix_u[jm]; 
                } 
                //Right border
                if( s_j == my && j < ny-1) // (s_i != BLOCK_SIZE_X && s_j == BLOCK_SIZE_Y && j != ny - 1)
                {
                    A_tile[s_i][my+1]=matrix_u[jp]; 
                } 
                // missing the corners. They are needed  only if using next nearest neighbors
                item.barrier(access::fence_space::local_space); // barrier within the group //__syncthreads();
                if(i>0 && i<nx-1 && j>0 && j< ny-1){
                    matrix_unew[ind] = factor * (A_tile[s_i-1][s_j]+A_tile[s_i-1][s_j]-2*A_tile[s_i][s_j])
                                    + factor * (A_tile[s_i][s_j+1]+A_tile[s_i][s_j-1]-2*A_tile[s_i][s_j]);
                }
                });
        });       
        q.wait();
        kernel_duration += (e.get_profiling_info<info::event_profiling::command_end>() - e.get_profiling_info<info::event_profiling::command_start>());
        //std::swap(matrix_u, matrix_unew);
    }

 
    std::cout << "Kernel Execution Time : " << kernel_duration / 1e+9 << " seconds" << "\n";

    auto duration = std::chrono::high_resolution_clock::now().time_since_epoch().count() - start;
    std::cout << "Compute Duration      : " << duration / 1e+9 << " seconds\n";

    //# Print Output
    if (PRINT_OUTPUT_MATRIX){
    } else {
        std::cout << " [0][0] = " << matrix_u[0] << "\n";
    }
}
``` 
I compile the code with 
``` 
syclcc -O2 --hipsycl-targets="omp;cuda:sm_80" code.cpp
```

And I run it with `./a.out -n 4096 -m 32`  (n is size in one direction and m the size of the group)

**Comment by illuhad** [comment]
> The case with buffers seems to be significantly faster than case with buffers

What is faster than what? :-)


Here are some general thoughts:

`malloc_shared` on CUDA just calls `cudaMallocManaged`. The performance of NVIDIA's unified memory performance is usually quite good. Keep in mind that such an alloation will migrate memory pages between host and device depending on where they are touched. This means that you can construct pathological examples where they are not efficient if both host and device touch the same page multiple times in quick succession, causing multiple data transfers. For real world cases this is mostly not an issue.

A buffer may have some overhead because it might manage additional allocations (e.g. a host allocation) in addition to the device allocation, and there is some small overhead associated with the automatic dependency analysis and task graph construction that happens in the buffer-accessor model. These overheads usually do not play a significant role, but you might notice them for very small kernels when the SYCL runtime's kernel submission latency becomes important.

So, in short: A significant difference is not expected for most cases on NVIDIA and Linux. I'm currently at a conference and don't have a lot of time to try to run your code at the moment, so apologies that I could not drill down more to what's going on.

**Comment by csccva** [comment]
Hello,

Thank you for your reply. This is a major facepalm for me. I spent lots of time doing tests and denigging and I did not noticed that the number of iterations was not the same. 
Once I fixed that the execution time was comparable as it should be. 

I apologize. I need a break from SYCL :) 
I wonder if there is some forum/slack space to put questions.

**Comment by illuhad** [comment]
No worries, this happens :)

For general SYCL questions, there is a KhronosDevs slack (but messages disappear after some time, so maybe not so useful for discussions that might also be interesting for others), or you can use the `sycl` tag on stackoverflow.

There is no hipSYCL-specific slack, but you can use github discussions on this repository for hipSYCL-specific questions, which is similar to a forum or slack: https://github.com/illuhad/hipSYCL/discussions

**Comment by csccva** [comment]
Thanks for the reply.  I will use the discussions section in the future.
